File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/bestia/ActivateBestiaActor.java
package net.bestia.zoneserver.actor.bestia;

import java.util.Objects;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import akka.actor.AbstractActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import net.bestia.messages.bestia.BestiaActivateMessage;
import net.bestia.zoneserver.actor.client.ClientMessageActor.RedirectMessage;
import net.bestia.zoneserver.entity.PlayerEntityService;

/**
 * Upon receiving an activation request from this account we check if the
 * account is able to uses this bestia. It will then get activated and all
 * needed information about the newly activated bestia is send to the client.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class ActivateBestiaActor extends AbstractActor {

	private final LoggingAdapter LOG = Logging.getLogger(getContext().system(), this);

	public final static String NAME = "activateBestia";

	private final PlayerEntityService playerService;

	@Autowired
	public ActivateBestiaActor(PlayerEntityService playerService) {

		this.playerService = Objects.requireNonNull(playerService);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder()
				.match(BestiaActivateMessage.class, this::handleActivateBestia)
				.build();
	}
	
	@Override
	public void preStart() throws Exception {
		final RedirectMessage msg = RedirectMessage.Companion.get(BestiaActivateMessage.class);
		context().parent().tell(msg, getSelf());
	}

	private void handleActivateBestia(BestiaActivateMessage msg) {
		
		final BestiaActivateMessage bestiaMsg = (BestiaActivateMessage) msg;

		try {
			playerService.setActiveEntity(bestiaMsg.getAccountId(), bestiaMsg.getEntityId());
			LOG.debug("Activated player bestia from accId: {}, entityId: {}",
					bestiaMsg.getAccountId(),
					bestiaMsg.getEntityId());

		} catch (IllegalArgumentException ex) {
			LOG.warning("Can not activiate entity: {}", bestiaMsg.toString());
		}
		
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/bestia/ActivateBestiaActor.kt
package net.bestia.zoneserver.actor.bestia

import mu.KotlinLogging
import net.bestia.messages.bestia.BestiaSetActive
import net.bestia.zoneserver.actor.Actor
import net.bestia.zoneserver.actor.entity.awaitEntityResponse
import net.bestia.zoneserver.actor.entity.component.UpdateComponent
import net.bestia.zoneserver.actor.routing.DynamicMessageRoutingActor
import net.bestia.zoneserver.actor.routing.MessageApi
import net.bestia.zoneserver.entity.PlayerEntityService
import net.bestia.zoneserver.entity.component.MetadataComponent

private val LOG = KotlinLogging.logger { }

/**
 * Upon receiving an activation request from this account we check if the
 * account is able to uses this bestia. It will then get activated and all
 * needed information about the newly activated bestia is send to the client.
 *
 * @author Thomas Felix
 */
@Actor
class ActivateBestiaActor(
    private val playerService: PlayerEntityService,
    private val messageApi: MessageApi
) : DynamicMessageRoutingActor() {

  override fun createReceive(builder: BuilderFacade) {
    builder.matchRedirect(BestiaSetActive::class.java, this::handleActivateBestia)
  }

  private fun handleActivateBestia(msg: BestiaSetActive) {
    playerService.getActivePlayerEntityId(msg.accountId)?.let { currentActiveId ->
      awaitEntityResponse(messageApi, context, currentActiveId) {
        val currentMetaData = it.tryGetComponent(MetadataComponent::class.java)
            ?: MetadataComponent(entityId = it.id)
        val updateMsg = UpdateComponent(
            currentMetaData.copyWithoutKey(MetadataComponent.PLAYER_IS_ACTIVE)
        )
        messageApi.send(updateMsg)
      }
    }

    awaitEntityResponse(messageApi, context, msg.entityId) {
      val currentMetaData = it.tryGetComponent(MetadataComponent::class.java)
          ?: MetadataComponent(entityId = it.id)
      val updatedComponent = UpdateComponent(
          currentMetaData.copyWith(MetadataComponent.PLAYER_IS_ACTIVE, true)
      )

      LOG.debug { "Activated Player Bestia from accId: ${msg.accountId}, entityId: ${msg.entityId}" }
      messageApi.send(updatedComponent)
    }
  }

  companion object {
    const val NAME = "activateBestia"
  }
}
